application-diagram-ui/src/main/resources/Diagram/DiagramEditSheet.xml

```diff
@@ -245,12 +245,36 @@ define('diagram-store', ['jquery', 'xwiki-meta', 'xwiki-utils', 'diagram-utils',
     left.done($.proxy(right, 'resolve')).fail($.proxy(right, 'reject'));
   };
 
+  var saveBlobAsImageAttachment = function(blob, fileName, documentReference) {
+    var deferred = $.Deferred();
+    var attachmentReference = new XWiki.AttachmentReference(fileName, documentReference);
+    // Avoid creating too many versions of the attachment.
+    xutils.deleteAttachment(attachmentReference).always(function() {
+      pipeDeferred(xutils.uploadAttachment(blob, attachmentReference), deferred);
+    });
+    return deferred.promise();
```

Comment:
``pipeDeferred`` is useful when you have to deal with callbacks, i.e. when the function you call doesn't return a promise but instead it accepts a callback. In this case the methods you call already return a promise so you can do simply:

    // We upload the attachment even if we failed to delete it first.
    return xutils.deleteAttachment(attachmentReference).then(uploadAttachment, uploadAttachment);

where ``uploadAttachment`` is defined before using ``$.proxy()``.